File: js/src/unicode.ts
/**
 * Unicode primitives with the semantics of the .NET APIs the matcher was written against.
 *
 * Every Unicode-dependent operation in the port goes through this module (lint enforces it).
 * Differences from JavaScript's defaults were measured against .NET 10 on every code point
 * (spec 003, research R1):
 *
 * - whitespace is .NET's `char.IsWhiteSpace` set, not `\s` (U+0085 is whitespace, U+FEFF is not);
 * - lower-casing is one UTF-16 unit at a time, and leaves a unit alone when JavaScript would expand
 *   it (U+0130 stays U+0130), so there is no final-sigma or length change;
 * - categories are per UTF-16 unit where .NET tests a `char`, and per code point where .NET uses
 *   `GetUnicodeCategory(string, index)`;
 * - NFKC runs between Unicode noncharacters, which pass through unchanged (research R2).
 *
 * @packageDocumentation
 */

/** A Unicode general category, as the two-letter code (`'Lu'`, `'Mn'`, …; `'Cn'` when unassigned). */
export type GeneralCategory =
  | 'Lu' | 'Ll' | 'Lt' | 'Lm' | 'Lo'
  | 'Mn' | 'Mc' | 'Me'
  | 'Nd' | 'Nl' | 'No'
  | 'Zs' | 'Zl' | 'Zp'
  | 'Cc' | 'Cf' | 'Cs' | 'Co' | 'Cn'
  | 'Pc' | 'Pd' | 'Ps' | 'Pe' | 'Pi' | 'Pf' | 'Po'
  | 'Sm' | 'Sc' | 'Sk' | 'So';

// Every category except 'Cn', which is what remains when none matches. Codes are position + 1, so a
// cache entry of 0 means "not computed yet".
const TESTED: readonly GeneralCategory[] = [
  'Lo', 'Ll', 'Lu', 'Mn', 'Po', 'Nd', 'Zs', 'So', 'Cf', 'Cc', 'Lm', 'Lt', 'Mc', 'Me', 'Nl', 'No',
  'Zl', 'Zp', 'Cs', 'Co', 'Pc', 'Pd', 'Ps', 'Pe', 'Pi', 'Pf', 'Sm', 'Sc', 'Sk',
];

const PATTERNS: readonly RegExp[] = TESTED.map(name => new RegExp(`^\\p{gc=${name}}$`, 'u'));

const UNASSIGNED = TESTED.length + 1;

let unitCache: Uint8Array | undefined;

function classify(text: string): number {
  for (let i = 0; i < PATTERNS.length; i++) {
    if (PATTERNS[i]!.test(text)) {
      return i + 1;
    }
  }

  return UNASSIGNED;
}

function nameOf(code: number): GeneralCategory {
  return code === UNASSIGNED ? 'Cn' : TESTED[code - 1]!;
}

/**
 * The general category of one UTF-16 code unit, like .NET `CharUnicodeInfo.GetUnicodeCategory(char)`.
 * A lone surrogate unit is `'Cs'`.
 */
export function categoryOfUnit(unit: number): GeneralCategory {
  const cache = (unitCache ??= new Uint8Array(0x10000));
  let code = cache[unit]!;
  if (code === 0) {
    code = classify(String.fromCharCode(unit));
    cache[unit] = code;
  }

  return nameOf(code);
}

/**
 * The general category of the code point at `index`, like .NET
 * `CharUnicodeInfo.GetUnicodeCategory(string, int)`: a valid surrogate pair is read as one code point.
 */
export function categoryAt(text: string, index: number): GeneralCategory {
  const unit = text.charCodeAt(index);
  if (isHighSurrogate(unit) && index + 1 < text.length && isLowSurrogate(text.charCodeAt(index + 1))) {
    return nameOf(classify(text.substring(index, index + 2)));
  }

  return categoryOfUnit(unit);
}

/** Whether a unit is a letter (L*), like .NET `char.IsLetter`. */
export function isLetter(unit: number): boolean {
  if (unit < 0x80) {
    return (unit >= 0x41 && unit <= 0x5a) || (unit >= 0x61 && unit <= 0x7a);
  }

  const category = categoryOfUnit(unit);
  return category === 'Lu' || category === 'Ll' || category === 'Lt' || category === 'Lm' || category === 'Lo';
}

/** Whether a unit is a letter (L*) or a decimal digit (Nd), like .NET `char.IsLetterOrDigit`. */
export function isLetterOrDigit(unit: number): boolean {
  if (unit < 0x80) {
    return (unit >= 0x30 && unit <= 0x39) || (unit >= 0x41 && unit <= 0x5a) || (unit >= 0x61 && unit <= 0x7a);
  }

  return isLetter(unit) || categoryOfUnit(unit) === 'Nd';
}

/** Whether a unit is a control character (Cc), like .NET `char.IsControl`. */
export function isControl(unit: number): boolean {
  return unit <= 0x1f || (unit >= 0x7f && unit <= 0x9f);
}

/**
 * Whether a unit is whitespace, like .NET `char.IsWhiteSpace`: categories Zs, Zl and Zp, plus
 * U+0009–U+000D, U+0085 and U+00A0. Unlike JavaScript's `\s`, U+0085 is whitespace and U+FEFF is not.
 */
export function isWhiteSpace(unit: number): boolean {
  if (unit < 0x80) {
    return unit === 0x20 || (unit >= 0x09 && unit <= 0x0d);
  }

  if (unit === 0x85 || unit === 0xa0) {
    return true;
  }

  const category = categoryOfUnit(unit);
  return category === 'Zs' || category === 'Zl' || category === 'Zp';
}

/** Whether a unit is a high or low surrogate, like .NET `char.IsSurrogate`. */
export function isSurrogate(unit: number): boolean {
  return unit >= 0xd800 && unit <= 0xdfff;
}

/** Whether a unit is a high (leading) surrogate, like .NET `char.IsHighSurrogate`. */
export function isHighSurrogate(unit: number): boolean {
  return unit >= 0xd800 && unit <= 0xdbff;
}

/** Whether a unit is a low (trailing) surrogate, like .NET `char.IsLowSurrogate`. */
export function isLowSurrogate(unit: number): boolean {
  return unit >= 0xdc00 && unit <= 0xdfff;
}

/**
 * Lower-cases one UTF-16 unit, like .NET `char.ToLowerInvariant`: when JavaScript's lower-casing
 * would not give exactly one unit (U+0130 `İ` gives two), the unit is returned unchanged.
 */
export function toLowerInvariant(unit: number): number {
  if (unit < 0x80) {
    return unit >= 0x41 && unit <= 0x5a ? unit + 0x20 : unit;
  }

  const lower = String.fromCharCode(unit).toLowerCase();
  return lower.length === 1 ? lower.charCodeAt(0) : unit;
}

/** Whether a code point is a Unicode noncharacter: U+FDD0–U+FDEF, or one ending in FFFE or FFFF. */
export function isNoncharacter(codePoint: number): boolean {
  return (codePoint >= 0xfdd0 && codePoint <= 0xfdef) || (codePoint & 0xfffe) === 0xfffe;
}

/**
 * The length (0, 1 or 2 units) of a noncharacter starting at `index`, or 0 when there is none.
 */
export function noncharacterLengthAt(text: string, index: number): number {
  const unit = text.charCodeAt(index);
  if (isHighSurrogate(unit) && index + 1 < text.length) {
    const low = text.charCodeAt(index + 1);
    if (isLowSurrogate(low)) {
      const codePoint = 0x10000 + ((unit - 0xd800) << 10) + (low - 0xdc00);
      return isNoncharacter(codePoint) ? 2 : 0;
    }
  }

  return isNoncharacter(unit) ? 1 : 0;
}

/**
 * NFKC, like .NET `string.Normalize(NormalizationForm.FormKC)` as PersianTextGuard applies it:
 * text between noncharacters is normalized and each noncharacter is copied unchanged. A noncharacter
 * composes with nothing, so the result equals whole-string NFKC. Lone surrogates must already have
 * been replaced by the caller.
 */
export function nfkc(text: string): string {
  let start = 0;
  let result = '';
  for (let i = 0; i < text.length; ) {
    const length = noncharacterLengthAt(text, i);
    if (length === 0) {
      i++;
      continue;
    }

    if (i > start) {
      result += text.substring(start, i).normalize('NFKC');
    }

    result += text.substring(i, i + length);
    i += length;
    start = i;
  }

  return start === 0 ? text.normalize('NFKC') : result + text.substring(start).normalize('NFKC');
}

/** Whether `text` is already in NFKC, like .NET `string.IsNormalized(NormalizationForm.FormKC)`. */
export function isNfkc(text: string): boolean {
  return nfkc(text) === text;
}

/** NFD, like .NET `string.Normalize(NormalizationForm.FormD)`. */
export function nfd(text: string): string {
  return text.normalize('NFD');
}

File: python/src/persian_text_guard/_unicode.py
"""Unicode primitives with the semantics of the .NET APIs the matcher was written against.

Port of ``js/src/unicode.ts``. Every Unicode-dependent operation in the package goes through this
module (``scripts/check_unicode_usage.py`` and ruff enforce it). The differences from Python's own
defaults were measured against .NET 10 on every code point (spec 004, research R1):

- whitespace is .NET's ``char.IsWhiteSpace`` set, not ``str.isspace``, which also counts
  U+001C-U+001F (U+0085 is whitespace, U+FEFF is not);
- lower-casing is one UTF-16 unit at a time, and leaves a unit alone when Python would expand it
  (U+0130 stays U+0130), so there is no final-sigma or length change;
- categories are per UTF-16 unit where .NET tests a ``char``, and per code point where .NET uses
  ``GetUnicodeCategory(string, index)``;
- NFKC runs between Unicode noncharacters, which pass through unchanged.

The functions take the text the matcher works on: the UTF-16 view from ``_utf16``, in which a
character above U+FFFF is two surrogate code points. A "unit" is a one-character ``str`` of that view.
"""

from __future__ import annotations

import re
import unicodedata

_LETTER_CATEGORIES = frozenset({"Lu", "Ll", "Lt", "Lm", "Lo"})

# .NET char.IsWhiteSpace: U+0009-U+000D, U+0085, and the Zs, Zl and Zp characters. Written out, because
# every category set here is identical on CPython 3.11-3.14 (Unicode 14-16), and a set lookup is the
# fastest test; tests/test_unicode.py checks it against unicodedata on the running interpreter.
WHITE_SPACE = frozenset(
    "\x09\x0a\x0b\x0c\x0d\x20\x85\xa0\U00001680"
    "\U00002000\U00002001\U00002002\U00002003\U00002004\U00002005\U00002006\U00002007\U00002008\U00002009\U0000200a"
    "\U00002028\U00002029\U0000202f\U0000205f\U00003000"
)

# One-unit categories, cached: every thread computes the same value, so a racing write is harmless
# (research R17). Units are UTF-16 units, so the cache holds at most 65,536 entries.
_categories: dict[str, str] = {}

_SURROGATE = re.compile("[\ud800-\udfff]")
_SURROGATE_PAIR = re.compile("[\ud800-\udbff][\udc00-\udfff]")
_SUPPLEMENTARY = re.compile("[\U00010000-\U0010ffff]")
_NONCHARACTER_BMP = re.compile("[\U0000fdd0-\U0000fdef\U0000fffe\U0000ffff]")


def category_of_unit(c: str) -> str:
    """The general category of one UTF-16 unit, like .NET ``GetUnicodeCategory(char)``.

    A lone surrogate unit is ``"Cs"``; an unassigned one is ``"Cn"``.
    """
    category = _categories.get(c)
    if category is None:
        category = _categories.setdefault(c, unicodedata.category(c))
    return category


def category_at(text: str, index: int) -> str:
    """The general category of the code point at ``index``, like .NET ``GetUnicodeCategory(string, int)``.

    A valid surrogate pair is read as one code point.
    """
    c = text[index]
    if "\ud800" <= c <= "\udbff" and index + 1 < len(text) and "\udc00" <= text[index + 1] <= "\udfff":
        return unicodedata.category(combine(c, text[index + 1]))
    return category_of_unit(c)


def combine(high: str, low: str) -> str:
    """The code point a high and a low surrogate unit encode, like .NET ``char.ConvertToUtf32``."""
    return chr(0x10000 + ((ord(high) - 0xD800) << 10) + (ord(low) - 0xDC00))


def is_letter(c: str) -> bool:
    """Whether a unit is a letter (L*), like .NET ``char.IsLetter``."""
    if c < "\x80":
        return "a" <= c <= "z" or "A" <= c <= "Z"
    return category_of_unit(c) in _LETTER_CATEGORIES


def is_letter_or_digit(c: str) -> bool:
    """Whether a unit is a letter (L*) or a decimal digit (Nd), like .NET ``char.IsLetterOrDigit``."""
    if c < "\x80":
        return "a" <= c <= "z" or "A" <= c <= "Z" or "0" <= c <= "9"
    category = category_of_unit(c)
    return category in _LETTER_CATEGORIES or category == "Nd"


def is_control(c: str) -> bool:
    """Whether a unit is a control character (Cc), like .NET ``char.IsControl``."""
    return c <= "\x1f" or "\x7f" <= c <= "\x9f"


def is_white_space(c: str) -> bool:
    """Whether a unit is whitespace, like .NET ``char.IsWhiteSpace``.

    Categories Zs, Zl and Zp, plus U+0009-U+000D, U+0085 and U+00A0. Unlike ``str.isspace``,
    U+001C-U+001F are not whitespace.
    """
    return c in WHITE_SPACE


def is_surrogate(c: str) -> bool:
    """Whether a unit is a high or low surrogate, like .NET ``char.IsSurrogate``."""
    return "\ud800" <= c <= "\udfff"


def is_high_surrogate(c: str) -> bool:
    """Whether a unit is a high (leading) surrogate, like .NET ``char.IsHighSurrogate``."""
    return "\ud800" <= c <= "\udbff"


def is_low_surrogate(c: str) -> bool:
    """Whether a unit is a low (trailing) surrogate, like .NET ``char.IsLowSurrogate``."""
    return "\udc00" <= c <= "\udfff"


def to_lower_invariant(c: str) -> str:
    """Lower-cases one UTF-16 unit, like .NET ``char.ToLowerInvariant``.

    When Python's lower-casing would not give exactly one character (U+0130 ``İ`` gives two), the
    unit is returned unchanged.
    """
    if c < "\x80":
        return c.lower()
    lower = c.lower()
    return lower if len(lower) == 1 else c


def is_noncharacter(code_point: int) -> bool:
    """Whether a code point is a Unicode noncharacter: U+FDD0-U+FDEF, or one ending in FFFE or FFFF."""
    return 0xFDD0 <= code_point <= 0xFDEF or (code_point & 0xFFFE) == 0xFFFE


def noncharacter_length_at(text: str, index: int) -> int:
    """The length (1 or 2 units) of a noncharacter starting at ``index``, or 0 when there is none.

    A surrogate pair in the view is read as one code point.
    """
    c = text[index]
    if "\ud800" <= c <= "\udbff" and index + 1 < len(text):
        low = text[index + 1]
        if "\udc00" <= low <= "\udfff":
            return 2 if is_noncharacter(ord(combine(c, low))) else 0
    return 1 if is_noncharacter(ord(c)) else 0


def recombine(text: str) -> str:
    """``text`` with every valid surrogate pair replaced by the code point it encodes.

    Lone surrogates are kept as they are.
    """
    if _SURROGATE.search(text) is None:
        return text
    return _SURROGATE_PAIR.sub(lambda m: combine(m.group()[0], m.group()[1]), text)


def split_supplementary(text: str) -> str:
    """``text`` with every code point above U+FFFF replaced by its two surrogate units."""
    if text.isascii() or _SUPPLEMENTARY.search(text) is None:
        return text
    return _SUPPLEMENTARY.sub(_split, text)


def _split(match: re.Match[str]) -> str:
    code_point = ord(match.group()) - 0x10000
    return chr(0xD800 + (code_point >> 10)) + chr(0xDC00 + (code_point & 0x3FF))


def nfkc(text: str) -> str:
    """NFKC, like .NET ``string.Normalize(NormalizationForm.FormKC)`` as PersianTextGuard applies it.

    ``text`` is a UTF-16 view, and so is the result: valid surrogate pairs are recombined into code
    points, normalized, and split again. Noncharacters are copied unchanged and normalized around.
    ``unicodedata`` keeps them, and a noncharacter composes with nothing, so this equals NFKC of the
    runs between them. Lone surrogates must already have been replaced by the caller.
    """
    if text.isascii():
        return text
    if _SURROGATE.search(text) is None:
        if _NONCHARACTER_BMP.search(text) is None:
            return unicodedata.normalize("NFKC", text)
        return _nfkc_around_noncharacters(text)
    return split_supplementary(_nfkc_around_noncharacters(recombine(text)))


def _nfkc_around_noncharacters(text: str) -> str:
    # text holds code points here, not a view.
    parts: list[str] = []
    start = 0
    for i, c in enumerate(text):
        if is_noncharacter(ord(c)):
            if i > start:
                parts.append(unicodedata.normalize("NFKC", text[start:i]))
            parts.append(c)
            start = i + 1
    if start == 0:
        return unicodedata.normalize("NFKC", text)
    parts.append(unicodedata.normalize("NFKC", text[start:]))
    return "".join(parts)


def is_nfkc(text: str) -> bool:
    """Whether ``text`` is already in NFKC, like .NET ``string.IsNormalized(NormalizationForm.FormKC)``."""
    if text.isascii():
        return True
    return unicodedata.is_normalized("NFKC", recombine(text))


def nfd(text: str) -> str:
    """NFD, like .NET ``string.Normalize(NormalizationForm.FormD)``, on a UTF-16 view."""
    return split_supplementary(unicodedata.normalize("NFD", recombine(text)))


def unicode_version() -> str:
    """The Unicode version of the running interpreter's character database."""
    return unicodedata.unidata_version


# A run of .NET whitespace, for collapsing it with one C-level pass.
WHITE_SPACE_RUN = re.compile("[" + "".join(re.escape(c) for c in sorted(WHITE_SPACE)) + "]+")
